### DISCUSSION TITLE
Page - Top, Bottom, Side, Controls, Status

### DISCUSSION THREAD

**Discussion by fredchristianson** [fredchristianson/javascript-game-engine#7] [Ideas]
In this image, everything with a blue background is controlled by the Game Application.  Everything with an orange background is controlled by the current Game.  
<img src='https://user-images.githubusercontent.com/3279104/223861634-1f46ce85-7510-4d61-b963-7a27e74723ef.png' title='Hello World Game' width='300'/>

What should the real layout be?  
- Top & bottom controlled by the app?
- Side bars?
- Phone, tablet, other size screens.

What controls and statuses do we want?
-Start/stop
-pause
-save/load
- high score
- win/loss counts
- time played.  this game and ever
- ???

**Comment by Eddie7145** [comment]
First of all could explain what the Game application means

**Reply by fredchristianson** [reply]
I'm not great about explaining things.  Ask questions any time.  My thinking is that there is a single page that handles all of the non-play stuff - that's what I'm calling the Game Application.  It will start/stop games, manage logging, users/accounts/servers in future versions if we add that type of functionality.  That's a decision we can change if we want each game to have full control.

After task-000-skeleton there is a page http://localhost:4263/app/index.html.  And a game "launcher" that lists other games and lets you pick one.  
The areas in blue here are controls that span games.  When you start a game, that game takes over all of the areas in orange.  Game-specific controls would go there.  And the Application controls can be hidden if we want full screen and the game has it's own controls for things like pause and quit.
![image](https://user-images.githubusercontent.com/3279104/224020661-af385366-d55c-4d05-998f-04fe139d30a8.png)

**Comment by Eddie7145** [comment]
For the layout in my own point of view, its better if the layout changes based on the different world they're transgressing into

And that way we would give them a more interesting and immersive experience

**Reply by fredchristianson** [reply]
You are right.  My thoughts are everything in orange in the image would be controlled by the game.  There may be some defaults.  And the blue (top & bottom bars) can be hidden.

**Comment by Eddie7145** [comment]
It would be preferred if it was the phone, tablets and other screen for now

Then later with little updates we could add for the app

**Reply by fredchristianson** [reply]
What I'm calling an app is an HTML page.  May need to work on my terminology.  Do we need a discussion or document for terms?

**Reply by RimuhRimu** [reply]
I agree that it might be helpful

**Comment by Eddie7145** [comment]
For the controls I feel like all of them are useful, it all just depends on the type of game each world is offering

**Comment by Eddie7145** [comment]
Yeah it would be nice is we had a leader board section, where highest scores by player will be displayed, so as to promote competition between players. 

Then we would also have some like goals for each world where when goals are meet it will progress to the next world.

Also to give users a bit of control, we would allow the to set goals for something in the game

**Reply by fredchristianson** [reply]
I'm using "World" as the game area.  Could be a background color and deck of cards, or pac-man, or a 3D landscape with moving targets to shoot.
So there's no moving to next world at the application level.  An individual game can have multiple worlds you can move to.  May need a better term than world.  
World has been the base class of the game area in projects I'd done before.  Pretty important so need to get the name right or make sure everyone understands the terminology.  Maybe "class GameArea" is better?

**Reply by fredchristianson** [reply]
I had thought of distributed play as version 2.  We can have 
- multiple players on same machine (first)
- multiple players on local network with same server (soon)
- multiple players on distributed network (internet) future.  May be a good project for AWS lambdas & stores (or Azure or wherever).  Local servers can use a lambda API for shared game stats (high scores, etc)

**Reply by chylinski82** [reply]
let me make sure I'm on the same page here as well, so if  there's no moving to next world at the application level and app is an HTML page, would that mean that every world, including in the same game, is a separate document/ web page?

**Reply by fredchristianson** [reply]
Only one web page the way I am thinking.  But areas that will be filled with different HTML & CSS depending on the game.  

The core javascript modules will always be available.  Additional modules will be loaded as needed for the game being played.

We're using a browser for UI but it's not really based on pages.  It's like a single-page web app.

I'm still testing to make sure there aren't issues, but plan to use [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM) to isolate the game elements/style from the core/app/page.

**Comment by chylinski82** [comment]
in terms of layout, sidebar sounds good on wide screen, definitely desktop, maybe tablet. Phone and tablet in portrait orientation no side bar, but like pop up semi transparent tiles with whatever would be otherwise on a sidebar, like perhaps the things from blue background controlled by the game application (save game, load game, exit, etc?). Game controls would feel natural in bottom right corner, but could be moved by the player. Game status, i vote on top in the center, alternatively on sidebar if there is one (might be confusing with the menu controlled by game app, but not necessarily, if we make it distinct enough).

What controls? depending on the game, probably arrows, for those who feel more comfortable controlling with mouse than keyboard, plus whatever needed, i.e. jump, skip turn, etc.

In terms of statuses controlled by the game: hi-score, lives, health, strength, point, etc. Again, depends on the game, i guess we would have them all in one Class/object and then disable some of them, depending what game is active.

**Reply by fredchristianson** [reply]
Great comment.  I hadn't thought of orientation or game controls.  
For statuses and other things we can have implementations of the common ones and games pick which ones they add to the collection.  Games can also create custom status, control, piece, etc.  
Games may also restrict what common controls are available.  I see quit and pause being at the app level instead of game, but some games may not allow pause.  Maybe other common controls like Save/load.

**Reply by chylinski82** [reply]
Thanks. Ok, so whatever the game uses, in terms of controls, statuses, options, etc, will be available in the app level for the game to add it to it's own collection. got it.
